app/models/main-graph.ts

```diff
@@ -0,0 +1,336 @@
+// File: main-graph.ts
+// Created by: Sheyne Anderson
+// Date created: January 22, 2017
+//
+//
+
+import { EventEmitter } from "@angular/core";
+import {
+    DrawableElement,
+    Drawable,
+    DrawableGraph,
+    DrawableEdge,
+    DrawableNode,
+    EdgeValidator,
+    DrawableEdgeEventArgs,
+    DrawableNodeEventArgs,
+    PropertyChangedEventArgs
+} from "../components/graph-editor/graph-editor.component";
+
+import { CoreModel, CoreElement, CoreElementKind, Plugin, validateEdge, ObjectType } from "sinap-core";
+import { DoubleMap } from "./double-map";
+
+/**
+ * Contains a reference to a `CoreElement` and a `Drawable`.
+ * 
+ * Calls to get/set will update both. It also conatins a 
+ * `proxy` field. Reads and writes to this will intelligently update 
+ * the core element and the drawable. For any of these uses, if 
+ * the backer object (core/drawable) references another core or drawable
+ * this will return a `BridgingProxy` in its place. Sets will intellegently
+ * sync the backends even if the set `v` is a `BridgingProxy`
+ */
+export class BridgingProxy {
+    proxy: { [a: string]: any };
+
+    constructor(public core: CoreElement, public drawable: Drawable, public graph: MainGraph) {
+        this.proxy = new Proxy(core.data, {
+            set: (data, k, v) => this.set(k, v),
+            get: (data, k) => this.get(k),
+        });
+    }
+
+    set(k: PropertyKey, v: any, updateDrawable = true) {
+        let drawableValue = v;
+        let coreValue = v;
+
+        // If v is a drawable element or a core element,
+        // then when it gets set in the various proxied data
+        // structures, set it differently. 
+
+        if (v instanceof BridgingProxy) {
```

Comment:
Could we change BridgingProxy, potentially adding in a second class, so that we don't have to do instanceof checks?